Update rmw_fastrtps_cpp and rmw_fastrtps_shared_cpp QDs to QL2. (#456)
ros2/rmw_fastrtps: rmw_fastrtps_cpp/QUALITY_DECLARATION.md

# `rmw_fastrtps_cpp` Quality Declaration

> The package `rmw_fastrtps_cpp` claims to be in the **Quality Level 2** category.

Reviewer: I was going to say this package can't be level 2 yet because FastDDS isn't (item 5.iii of REP-2004), but it looks like it is!

https://github.com/eProsima/Fast-DDS/pull/1400/

### Feature Testing [4.i]

All `rmw_fastrtps_cpp` public features are ROS middleware features.

> Thus, unit, integration, and system tests up the stack, such as those found in [`test_rmw_implementation`](https://github.com/ros2/rmw_implementation/tree/master/test_rmw_implementation), [`test_rclcpp`](https://github.com/ros2/system_tests/tree/master/test_rclcpp), and [`test_communication`](https://github.com/ros2/system_tests/tree/master/test_communication) packages, provide feature coverage. Nightly CI jobs in [`ci.ros2.org`](https://ci.ros2.org/) and [`build.ros2.org`](https://build.ros2.org/), where `rmw_fastrtps_cpp` is the default `rmw` implementation, thoroughly exercise this middleware.

Reviewer 2: ```suggestion
Unit, integration, and system tests higher up in the stack, such as those found in [`test_rmw_implementation`](https://github.com/ros2/rmw_implementation/tree/master/test_rmw_implementation), [`test_rclcpp`](https://github.com/ros2/system_tests/tree/master/test_rclcpp), and [`test_communication`](https://github.com/ros2/system_tests/tree/master/test_communication) packages, provide feature coverage. Nightly CI jobs in [`ci.ros2.org`](https://ci.ros2.org/) and [`build.ros2.org`](https://build.ros2.org/), where `rmw_fastrtps_cpp` is the default `rmw` implementation, thoroughly exercise this middleware.
```

Reviewer 2: (same comment for below)

Author: Sounds good, and it applies to `rmw_fastrtps_shared_cpp` as well. See https://github.com/ros2/rmw_fastrtps/pull/456/commits/f035c81af14b2fde6d8761e3748bd7037a301851.

### Public API Testing [4.ii]

> `rmw_fastrtps_cpp` implements ROS middleware public API, but also provides public API of its own.

Reviewer 2: ```suggestion
`rmw_fastrtps_cpp` implements the ROS middleware public API, but also provides public API of its own.
```

Reviewer 2: (same comment for below)

Author: See https://github.com/ros2/rmw_fastrtps/pull/456/commits/f035c81af14b2fde6d8761e3748bd7037a301851.

Changes are required to make a best effort to keep or increase coverage before being accepted, but decreases are allowed if properly justified and accepted by reviewers.

> This package does not meet the 95% coverage guideline, as error handling behavior is currently not being tested on system resource exhaustion and on `fastrtps` internal failure.

Author: Justification probably needs work.

> 
> Changes are required to make a best effort to keep or increase coverage before being accepted, but decreases are allowed if properly justified and accepted by reviewers.
> 
> This package does not meet the 95% coverage guideline, as error handling behavior is currently not being tested on system resource exhaustion and on `fastrtps` internal failure.
> Current coverage statistics can be viewed [here](https://ci.ros2.org/job/ci_linux_coverage/lastSuccessfulBuild/cobertura/src_ros2_rmw_fastrtps_rmw_fastrtps_cpp_src/).

Reviewer 2: I'd rearrange and reword this.

```suggestion
Current coverage statistics can be viewed [here](https://ci.ros2.org/job/ci_linux_coverage/lastSuccessfulBuild/cobertura/src_ros2_rmw_fastrtps_rmw_fastrtps_cpp_src/).

This package claims to meet the coverage requirements for the current quality level, even though it doesn't have 95% line coverage.  The justification is that the only uncovered lines have to do with system resource exhaustion and Fast-DDS internal failure.
```

Reviewer 2: (same comment for below)

Author: Suggestion accepted. See https://github.com/ros2/rmw_fastrtps/pull/456/commits/f035c81af14b2fde6d8761e3748bd7037a301851.
